Preserve sync subtree for '***'.

michaelfig opened the pull request:
Fixes #1807 

This PR implements subtree syncing for '***' globs in the sync specification.

Given:
```yaml
    sync:
      'foo/***/*.js': /usr/src/app/
```

and changing `foo/baz/bot/test.js`, `test.js` will be synchronized to `/usr/src/app/baz/bot/test.js`.

Given:
```yaml
    sync:
      '***': subdir
```

and changing `my/src.txt`, `src.txt` will be synchronized to `${WORKDIR}/subdir/my/src.txt`.

All the existing sync specs in 0.25.0 mean the same as they did before this PR (i.e. they collapse the subPath, so all files in the subtree are synced to the same destination directory, not its children).  To allow for backward compatibility with 0.23.0's subtree syncing, you can use:

```yaml
    sync:
      '**': /usr/src/app/ # v0.23.0
      '***': /usr/src/app/ # this PR
```

The line marked `this PR` will take precedence over the `v0.23.0` line (triple-stars are matched first before other patterns), except when running a version of Skaffold before this PR.  In that case, the `v0.23.0` line will be used, and only `v0.23.0` will produce the same behaviour as this PR.

Hope this helps,
Michael.

michaelfig commented on pkg/skaffold/sync/sync.go:
```diff
@@ -128,21 +129,43 @@ func intersect(context string, syncMap map[string]string, files []string, workin
+					if isTripleStar {
+						// Map the paths as a tree from the prefix.
+						subtreePrefix := strings.Split(p, "***")[0]
+						subPath = strings.TrimPrefix(relPath, subtreePrefix)
```
Need:
```go
filepath.FromSlash(subtreePrefix)
```
Otherwise the prefix-pruning fails on Windows.

tejal29 commented on pkg/skaffold/sync/sync_test.go:
```diff
@@ -275,6 +275,61 @@ func TestNewSyncItem(t *testing.T) {
+				},
+			},
+			evt: watch.Events{
+				Added: []string{filepath.Join("dir1", "dir2/node.js")},
```
i dont like the fact that this for loop is so long and it gets difficult to read.
How about?
```

tripleStarSyncMaps, otherSyncMaps = segregateSyncMaps(syncMaps)
for _, f := range files {
 ...
 var match bool
 // first try all tripleStarSyncMaps
 match, dst, err := matchesTripleStarSyncMap(f)
 if err != nil {
    return nil, err
 }
 if match {
   ret[f] = dst
   continue
 }
 // Try matching others
 match, dst, err := match(otherSyncMaps, f)
 if !match {
  logrus.Infof("Changed file %s does not match any sync pattern. Skipping sync", relPath)
  return nil, nil
 }
 ret[f] = dst
}
...
}


func segregateSyncMaps(syncMap map[string]string) (map[string]string, map[string] string) {
  for p, dst := range syncMap {
     pat := strings.Replace(p, "***", "**", -1)
     if len(pat) != len(p) {
       tripleStarPatter[pat] = dst
     } else {
      doubelStarPattern[p] = dst
    } 
 } 
 return tripleStarPattern, doubleStarPatter
}
```

That ways, we individually test smaller functions.

michaelfig replied:
AFAICT, the file watcher only returns relative events.  What, specifically, are you suggesting should be the absolute path?

michaelfig replied:
No, I see you're right.  I'll have a look.

tejal29 commented on pkg/skaffold/sync/sync.go:
```diff
@@ -128,21 +129,43 @@ func intersect(context string, syncMap map[string]string, files []string, workin
 			return nil, errors.Wrapf(err, "changed file %s can't be found relative to context %s", f, context)
 		}
 		var matches bool
-		for p, dst := range syncMap {
```
i dont like the fact that this for loop is so long and it gets difficult to read.
How about we first find all the TripleStar Patterns, match them and then try all the others?

```

tripleStarSyncMaps, otherSyncMaps = segregateSyncMaps(syncMaps)
for _, f := range files {
 ...
 var match bool
 // first try all tripleStarSyncMaps
 match, dst, err := matchesTripleStarSyncMap(f)
 if err != nil {
    return nil, err
 }
 if match {
   ret[f] = dst
   continue
 }
 // Try matching others
 match, dst, err := matchOthers(otherSyncMaps, f)
 if !match {
  logrus.Infof("Changed file %s does not match any sync pattern. Skipping sync", relPath)
  return nil, nil
 }
 ret[f] = dst
}
...
}


func segregateSyncMaps(syncMap map[string]string) (map[string]string, map[string] string) {
  for p, dst := range syncMap {
     pat := strings.Replace(p, "***", "**", -1)
     if len(pat) != len(p) {
       tripleStarPatter[pat] = dst
     } else {
      doubelStarPattern[p] = dst
    } 
 } 
 return tripleStarPattern, doubleStarPatter
}
```

That ways, we individually test smaller functions.

michaelfig replied:
Good suggestion.  Thanks!

tejal29 commented:
Thank you for the bug fix! The kokoro job failed due to intergration test flakiness #1538. 
This should be fixed. I have the kokoro-run label. Hopefully a new build will be kicked off soon.

michaelfig commented:
@tejal29 I implemented your proposed changes.  I found and fixed some test failures with absolute paths on Windows.  Would you mind please relabeling `kokoro-run`?

tejal29 commented on pkg/skaffold/sync/sync.go:
```diff
@@ -119,41 +120,108 @@ func latestTag(image string, builds []build.Artifact) string {
+			doubleStarPattern[p] = dst
+		}
+	}
+	return
```
Thanks, i never knew you can return like this in go.
